common/src/main/java/net/william278/husksync/database/MongoDbDatabase.java

```diff
@@ -0,0 +1,363 @@
+import net.william278.husksync.database.mongo.MongoCollectionHelper;
+import net.william278.husksync.database.mongo.MongoConnectionHandler;
+import net.william278.husksync.user.User;
+import org.bson.Document;
+import org.bson.conversions.Bson;
+import org.bson.types.Binary;
+import org.jetbrains.annotations.Blocking;
+import org.jetbrains.annotations.NotNull;
+
+import java.time.Instant;
+import java.time.OffsetDateTime;
+import java.util.List;
+import java.util.Optional;
+import java.util.TimeZone;
+import java.util.UUID;
+import java.util.logging.Level;
+
+public class MongoDbDatabase extends Database {
+    private MongoConnectionHandler mongoConnectionHandler;
+    private MongoCollectionHelper mongoCollectionHelper;
+
+    private final String usersTable;
+    private final String userDataTable;
+    public MongoDbDatabase(@NotNull HuskSync plugin) {
+        super(plugin);
+        this.usersTable = plugin.getSettings().getDatabase().getTableName(TableName.USERS);
+        this.userDataTable = plugin.getSettings().getDatabase().getTableName(TableName.USER_DATA);
+    }
+
+    /**
+     * Initialize the database and ensure tables are present; create tables if they do not exist.
+     *
+     * @throws IllegalStateException if the database could not be initialized
+     */
+    @Override
+    public void initialize() throws IllegalStateException {
+        final Settings.DatabaseSettings.DatabaseCredentials credentials = plugin.getSettings().getDatabase().getCredentials();
+        try {
+            mongoConnectionHandler = new MongoConnectionHandler(
+                    credentials.getHost(),
+                    credentials.getPort(),
+                    credentials.getUsername(),
+                    credentials.getPassword(),
+                    credentials.getDatabase(),
+                    credentials.getMongoAuthDb()
+            );
+            mongoCollectionHelper = new MongoCollectionHelper(mongoConnectionHandler);
+            if (mongoCollectionHelper.getCollection(usersTable) == null) {
+                mongoCollectionHelper.createCollection(usersTable);
+            }
+            if (mongoCollectionHelper.getCollection(userDataTable) == null) {
+                mongoCollectionHelper.createCollection(userDataTable);
+            }
+        } catch (Exception e) {
+            throw new IllegalStateException("Failed to establish a connection to the MongoDB database. " +
+                    "Please check the supplied database credentials in the config file", e);
+        }
+    }
+
+    /**
+     * Ensure a {@link User} has an entry in the database and that their username is up-to-date
+     *
+     * @param user The {@link User} to ensure
+     */
+    @Blocking
+    @Override
+    public void ensureUser(@NotNull User user) {
+        getUser(user.getUuid()).ifPresentOrElse(
+                existingUser -> {
+                    if (!existingUser.getUsername().equals(user.getUsername())) {
```

Comment:
you could do `if (existingUser.getUsername().equals(user.getUsername())) {
                            return;
                        }`

Reply:
the check logic is ripped from mysql impl

Reply:
thats just copied from the mysql impl